libeos-payg/fake-clock.c

```diff
@@ -0,0 +1,147 @@
+ * the terms of the MPL, indicate your decision by deleting the provisions
+ * above and replace them with the notice and other provisions required by the
+ * LGPL. If you do not delete the provisions above, a recipient may use your
+ * version of this file under the terms of either the MPL or the LGPL.
+ */
+
+#include <libeos-payg/fake-clock.h>
+#include <libeos-payg/boottime-source.h>
+
+#define MSEC_PER_SEC 1000
+
+struct _EpgFakeClock
+{
+  GObject parent_instance;
+
+  gint64 time; /* in microseconds */
+  gint64 wallclock_time; /* in microseconds */
+};
+
+static gint64
+epg_fake_clock_get_wallclock_time (EpgClock *clock)
+{
+  EpgFakeClock *self = (EpgFakeClock *)clock;
+
+  g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), -1);
+
+  return self->wallclock_time;
+}
+
+/**
+ * epg_fake_clock_set_wallclock_time:
+ * @self: an #EpgFakeClock
+ * @new_time: a nonnegative value to set the clock to
+ *
+ * Set the time to @new_time so that value will be returned by
+ * epg_clock_get_wallclock_time() until the next call to this function. This is
+ * intended to be used by unit tests.
+ *
+ * Since: 0.2.1
+ */
+void
+epg_fake_clock_set_wallclock_time (EpgFakeClock *self,
+                                   gint64        new_time)
+{
+  g_return_if_fail (EPG_IS_FAKE_CLOCK (self));
+  g_return_if_fail (new_time >= 0);
+
+  self->wallclock_time = new_time;
+}
+
+static gint64
+epg_fake_clock_get_time (EpgClock *clock)
+{
+  EpgFakeClock *self = (EpgFakeClock *)clock;
+
+  g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), -1);
+
+  return self->time;
+}
+
+/**
+ * epg_fake_clock_set_time:
+ * @self: an #EpgFakeClock
+ * @new_time: a nonnegative value to set the clock to
+ *
+ * Set the time to @new_time so that value will be returned by
+ * epg_clock_get_time() until the next call to this function. This is intended
+ * to be used by unit tests.
+ *
+ * Since: 0.2.1
+ */
+void
+epg_fake_clock_set_time (EpgFakeClock *self,
+                         gint64       new_time)
+{
+  g_return_if_fail (EPG_IS_FAKE_CLOCK (self));
+  g_return_if_fail (new_time >= 0);
+
+  self->time = new_time;
+}
+
+static GSource *
+epg_fake_clock_source_new_seconds (EpgClock *clock,
+                                   guint     interval)
+{
+  EpgFakeClock *self = (EpgFakeClock *)clock;
+  guint interval_clamped;
+
+  g_return_val_if_fail (EPG_IS_FAKE_CLOCK (self), NULL);
+
+  interval_clamped = MIN (interval, G_MAXUINT / MSEC_PER_SEC);
+  return epg_boottime_source_new (interval_clamped * MSEC_PER_SEC);
+}
+
+static void
+clock_iface_init (EpgClockInterface *iface)
+{
+  iface->get_wallclock_time = epg_fake_clock_get_wallclock_time;
+  iface->get_time = epg_fake_clock_get_time;
+  iface->source_new_seconds = epg_fake_clock_source_new_seconds;
+}
+
+G_DEFINE_TYPE_WITH_CODE (EpgFakeClock, epg_fake_clock, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EPG_TYPE_FAKE_CLOCK, clock_iface_init))
```

Comment:
This should be `EPG_TYPE_CLOCK`, not `EPG_TYPE_FAKE_CLOCK`. That’s what was causing the `epg_fake_clock_get_type()` recursion and hence a deadlock:
```
Thread 1 (Thread 0x7ffff6ba0380 (LWP 2587)):
#0  0x00007ffff751aead in syscall () from /lib64/libc.so.6
#1  0x00007ffff76c7160 in g_cond_wait (cond=0x7ffff7949560 <g_once_cond>, mutex=0x7ffff7949550 <g_once_mutex>) at /opt/endless/source/glib/glib/gthread-posix.c:1395
#2  0x00007ffff769fab8 in g_once_init_enter (location=0x7ffff7fc39d0 <g_define_type_id.volatile>) at /opt/endless/source/glib/glib/gthread.c:665
#3  0x00007ffff7f9e51e in epg_fake_clock_get_type () at ../../source/eos-payg/libeos-payg/fake-clock.c:116
#4  0x00007ffff7f9e4aa in epg_fake_clock_get_type_once () at ../../source/eos-payg/libeos-payg/fake-clock.c:116
#5  0x00007ffff7f9e552 in epg_fake_clock_get_type () at ../../source/eos-payg/libeos-payg/fake-clock.c:116
#6  0x00007ffff7f9e5f4 in epg_fake_clock_new () at ../../source/eos-payg/libeos-payg/fake-clock.c:146
#7  0x000000000040322e in manager_new_failable (fixture=0x41ee20, enabled=1, error=0x7fffffffc820) at ../../source/eos-payg/libeos-payg/tests/manager.c:166
#8  0x0000000000403368 in manager_new (fixture=0x41ee20) at ../../source/eos-payg/libeos-payg/tests/manager.c:181
#9  0x000000000040396b in test_manager_load_empty (fixture=0x41ee20, data=0x0) at ../../source/eos-payg/libeos-payg/tests/manager.c:241
#10 0x00007ffff769cb1f in test_case_run (tc=0x41fa90) at /opt/endless/source/glib/glib/gtestutils.c:2163
#11 0x00007ffff769cedf in g_test_run_suite_internal (suite=0x41f040, path=0x0) at /opt/endless/source/glib/glib/gtestutils.c:2247
#12 0x00007ffff769cf88 in g_test_run_suite_internal (suite=0x41f020, path=0x0) at /opt/endless/source/glib/glib/gtestutils.c:2259
#13 0x00007ffff769d19f in g_test_run_suite (suite=0x41f020) at /opt/endless/source/glib/glib/gtestutils.c:2335
#14 0x00007ffff769be2e in g_test_run () at /opt/endless/source/glib/glib/gtestutils.c:1600
#15 0x0000000000406b73 in main (argc=1, argv=0x7fffffffcc68) at ../../source/eos-payg/libeos-payg/tests/manager.c:789
```

Reply:
That makes sense; I think it was a replace-all error. Thanks for pointing it out!

Reply:
Called it :)

Reply:
🖁